store/store.go

```diff
@@ -168,23 +186,118 @@ func (s *Store) Commit() (root []byte, err lib.ErrorI) {
 	return bytes.Clone(s.root), nil
 }
 
+// PARTITIONING CODE BELOW
+
+// ShouldPartition() determines if it is time to partition
+func (s *Store) ShouldPartition() bool {
+	// check 10 times per partition range (in case there's a failure)
+	checkFrequency := partitionFrequency / 5
+	// check if it's time to partition
+	if s.version < partitionFrequency || s.version%checkFrequency != 1 {
+		return false
+	}
+	// get the partition exists value from the store at a particular historical partition prefix
+	value, err := s.hss.Get([]byte(partitionExistsKey))
+	if err != nil {
+		s.log.Errorf("ShouldPartition() failed with err: %s", err.Error())
+		return false
+	}
+	// check if the value is set <key is the value>
+	return !bytes.Equal(value, []byte(partitionExistsKey))
+}
+
+// Partition()
+//  1. SNAPSHOT: for each key in the state store, copy them in the new historical partition to ensure
+//     each partition has a complete snapshot and allows older partitions to safely be pruned
+//  2. PRUNE: Drop all versions in the LSS older than the latest @ partition height
+func (s *Store) Partition() {
+	s.log.Info("Started partition process ✂️")
+	if err := func() lib.ErrorI {
+		// calculate the partition height
+		partHeight := partitionHeight(s.version)
+		// create a writer at the partition height
+		reader := s.db.NewTransactionAt(partHeight, false)
+		defer reader.Discard()
+		// create a managed batch to do the 'writing'
+		writer := s.db.NewManagedWriteBatch()
+		defer writer.Cancel()
+		// generate the historical partition prefix
+		partitionPrefix := historicalPrefix(partHeight)
+		// get the latest store in a transaction wrapper
+		lss := NewTxnWrapper(reader, s.log, stateStorePrefix)
+		// create an iterator that traverses the entire state at the partition height
+		it, err := lss.Iterator(nil)
+		if err != nil {
+			return err
+		}
+		defer it.Close()
+		// set a signal the partition was successfully created <only written if batch succeeds>
+		if e := writer.SetEntryAt(&badger.Entry{Key: []byte(partitionPrefix + partitionExistsKey), Value: []byte(partitionExistsKey)}, partHeight); e != nil {
+			return ErrSetBatch(e)
+		}
+		// for each key in the state at the partition height
+		for ; it.Valid(); it.Next() {
+			// get the key from the iterator item
+			k, v := it.Key(), it.Value()
+			// set in the historical partition
+			if e := writer.SetEntryAt(&badger.Entry{Key: append([]byte(partitionPrefix), k...), Value: v}, partHeight); e != nil {
+				return ErrSetBatch(e)
+			}
+			// set in the latest store prefix with telling badgerDB that it may discard earlier versions
+			if e := writer.SetEntryAt((&badger.Entry{Key: append([]byte(stateStorePrefix), k...), Value: v}).WithDiscard(), partHeight); e != nil {
+				return ErrSetBatch(e)
+			}
+		}
+		// commit the batch
+		if e := writer.Flush(); e != nil {
+			return ErrFlushBatch(e)
+		}
+		// trigger garbage collector to prune keys
+		if e := s.db.RunValueLogGC(.25); e != nil {
```

Comment:
nitpick: `RunValueLogGC` usage should be protected with as it states on its description that one of them should run at a time, although is not useful now, adding this would be useful for when adding asynchronous partitioning during syncing or at least a comment for it